### PR TITLE
[EarlyCSE] De-Duplicate callsites with differing attrs

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[EarlyCSE] Add tests for de-duplication of callsites with differing attrs; NFC**
- **[EarlyCSE] De-Duplicate callsites with differing attrs**

We only do this if the attributes of the two callsites are compatible
(intersectable) which is probably not in fact necessary.

### CODE REVIEW

**Comment by goldsteinn** [comment]
AFAICT, we can just entirely ignore the attributes, but maybe I'm missing something.

File: llvm/lib/Transforms/Scalar/EarlyCSE.cpp [hunk 1]
```diff
@@ -1632,6 +1632,9 @@ bool EarlyCSE::processNode(DomTreeNode *Node) {
           LLVM_DEBUG(dbgs() << "Skipping due to debug counter\n");
           continue;
         }
+
+        // Potential TODO: We may be throwing away attribute information when
+        // we delete Inst that we could propagate too InVal.first.
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Scalar/EarlyCSE.cpp:1637]: I suspect we will run into correctness issues if we don't combine the attributes after CSE. But I cannot think of a counterexample right now.

```suggestion
        // we delete Inst that we could propagate to InVal.first.
```

Comment (goldsteinn) [reply llvm/lib/Transforms/Scalar/EarlyCSE.cpp:1637]: If we dont combine?

I was thinking about it, are there any **callsite** attributes that can create side-effects?
AFAICT no.

File: llvm/test/Transforms/EarlyCSE/replace-calls-def-attrs.ll [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+declare i8 @buz(i8, i8)
+define i8 @same_parent_combine_diff_attrs(i8 %x, i8 %y) {
+; CHECK-LABEL: define i8 @same_parent_combine_diff_attrs(
+; CHECK-SAME: i8 [[X:%.*]], i8 [[Y:%.*]]) {
+; CHECK-NEXT:    [[C1:%.*]] = call i8 @baz(i8 [[X]], i8 noundef [[Y]])
+; CHECK-NEXT:    [[R:%.*]] = call i8 @buz(i8 [[C1]], i8 [[C1]])
+; CHECK-NEXT:    ret i8 [[R]]
+;
+  %c1 = call i8 @baz(i8 %x, i8 noundef %y)
+  %c0 = call i8 @baz(i8 noundef %x, i8 noundef %y)
```
Comment (DianQK) [inline llvm/test/Transforms/EarlyCSE/replace-calls-def-attrs.ll:14]: Could you create a test case by swapping these two `baz`?

File: llvm/lib/Transforms/Scalar/EarlyCSE.cpp (hunk 1 above)
Comment (nikic) [reply llvm/lib/Transforms/Scalar/EarlyCSE.cpp:1637]: We still need to intersect attributes here. If say the first call returns `nonnull` but is (dynamically) unused, and the second one does not return `nonnull` but is used, you'll now replace it with a `nonnull` call and a potential poison value.

Comment (goldsteinn) [reply llvm/lib/Transforms/Scalar/EarlyCSE.cpp:1637]: Okay, I can see that argument for return values, but for param or fn?

Comment (goldsteinn) [reply llvm/lib/Transforms/Scalar/EarlyCSE.cpp:1637]: If we did CSE for non-readonly funtions ofc I could see param args mattering, but with the limits I think its only return?

File: llvm/include/llvm/IR/Instruction.h [hunk 3]
```diff
@@ -876,7 +876,8 @@ class Instruction : public User,
   /// Return true if the specified instruction is exactly identical to the
   /// current one. This means that all operands match and any extra information
   /// (e.g. load is volatile) agree.
-  bool isIdenticalTo(const Instruction *I) const LLVM_READONLY;
+  bool isIdenticalTo(const Instruction *I,
+                     bool IntersectAttrs = false) const LLVM_READONLY;
```
[the remark this answers is not in the document]
Comment (goldsteinn) [reply llvm/include/llvm/IR/Instruction.h:880]: Do we not need the `SubclassOptionalData == I->SubclassOptionalData`?

[the remark this answers is not in the document]
Comment (nikic) [reply llvm/include/llvm/IR/Instruction.h:880]: We indeed don't want to compare SubclassOptionalData -- though now that I double check the code, we do need a call to combineIRFlags() in the CallValue case. This should be testable for a call with FMF flags.

File: llvm/lib/Transforms/Scalar/EarlyCSE.cpp [hunk 4]
```diff
@@ -1632,6 +1639,18 @@ bool EarlyCSE::processNode(DomTreeNode *Node) {
           LLVM_DEBUG(dbgs() << "Skipping due to debug counter\n");
           continue;
         }
+        // We probably don't need a complete itersection of attrs between
+        // InVal.first and Inst.
+        // We only CSE readonly functions that have the same memory state and
+        // its not clear that any non-return *callsite* attribute can create
+        // side-effects. Likewise this implies when checking equality of
+        // callsite for CSEing, we can probably ignore all non-return attrs.
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/EarlyCSE.cpp:1647]: This is not right, there is really no distinction between return and parameter attributes here. Just move the attribute to the parameter and have the function return it.

There *is* a distinction between attributes that imply poison and those that imply UB, in that the latter can be unconditionally preserved from the first call-site, while the former need intersection. But just intersecting everything is conservatively correct.

File: llvm/include/llvm/IR/Instruction.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (nikic) [reply llvm/include/llvm/IR/Instruction.h:880]: And in that case I'd also move the attribute intersection into combineIRFlags.

File: llvm/lib/Transforms/Scalar/EarlyCSE.cpp (hunk 4 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/Scalar/EarlyCSE.cpp:1647]: Your argument is something like:

```
declare void @biz(ptr, ptr)
define ptr @bar(ptr %p) {
   ret ptr %p
}

define void @buz(ptr %p) {
  %p0 = call ptr @bar(ptr nonnull %p) readnone
  %p1 = call ptr @bar(ptr %p) readnone
  call void @biz(ptr %p0, ptr %p1)
}
```

Where we later propagate the `nonnull` to the return attr of `bar`?

Comment (goldsteinn) [reply llvm/lib/Transforms/Scalar/EarlyCSE.cpp]: Cleared up the comment.

**Review by dtcxzyw — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Scalar/EarlyCSE.cpp [hunk 5]
```diff
@@ -1307,6 +1307,23 @@ static void combineIRFlags(Instruction &From, Value *To) {
...
   }
+  if (isa<CallBase>(&From) && isa<CallBase>(To)) {
+    // NB: Intersection of attrs between InVal.first and Inst is overly
+    // conservative. Since we only CSE readonly functions that have the same
+    // memory state, we can preserve (or possibly in some cases combine)
+    // more attributes. Likewise this implies when checking equality of
+    // callsite for CSEing, we can probably ignore more attributes.
+    // Generally poison generating attributes need to be handled with more
+    // care as they can create *new* UB if preserved/combined and violated.
+    // Attributes that imply immediate UB on the otherhand would have been
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Scalar/EarlyCSE.cpp:1318]:
```suggestion
    // Attributes that imply immediate UB on the other hand would have been
```

Comment (goldsteinn) [reply llvm/lib/Transforms/Scalar/EarlyCSE.cpp:1318]: Ill change before pushing.

File: llvm/lib/Transforms/Scalar/EarlyCSE.cpp [hunk 6]
```diff
@@ -1307,6 +1307,23 @@ static void combineIRFlags(Instruction &From, Value *To) {
...
+    // violated either way.
+    bool Success =
+        cast<CallBase>(To)->tryIntersectAttributes(cast<CallBase>(&From));
+    assert(Success && "Failed to intersect attributes in callsites that "
+                      "passed identical check");
+    // For NDEBUG Compile.
+    (void)Success;
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/EarlyCSE.cpp:1325]: I'd use `[[maybe_unused]]` here, but either way is fine.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/Transforms/Scalar/EarlyCSE.cpp (hunk 6 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/Scalar/EarlyCSE.cpp:1325]: I tend to prefer the `(void)X`. Although not religiously.

**Pull request closed without merging**